File: app/models/db/learning_deck.py
from __future__ import annotations

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Index, Integer, String, Text, text
from sqlalchemy.dialects.postgresql import JSONB

from app.core.db import Base
from app.models.contracts import LearningDeckRunStatus
from app.models.db.common import _utcnow

ACTIVE_LEARNING_DECK_RUN_STATUS_VALUES = (
    LearningDeckRunStatus.QUEUED.value,
    LearningDeckRunStatus.PREPARING.value,
    LearningDeckRunStatus.GENERATING.value,
    LearningDeckRunStatus.VALIDATING.value,
    LearningDeckRunStatus.PUBLISHING.value,
)
ACTIVE_LEARNING_DECK_RUN_STATUS_SQL = (
    "status IN ('queued', 'preparing', 'generating', 'validating', 'publishing')"
)


class LearningDeck(Base):
    """Stable user-facing Learning Deck record for one source."""

    __tablename__ = "learning_decks"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, nullable=False, index=True)
    source_kind = Column(String(32), nullable=False, index=True)
    source_identity = Column(String(512), nullable=False)
    source_url = Column(String(2048), nullable=True)
    source_content_id = Column(Integer, nullable=True, index=True)
    source_title = Column(String(500), nullable=True)
    source_metadata = Column(JSONB, nullable=False, default=dict)
    title = Column(String(500), nullable=False)
    latest_successful_run_id = Column(Integer, nullable=True, index=True)
    latest_run_id = Column(Integer, nullable=True, index=True)
    artifact_storage_prefix = Column(String(2048), nullable=True)
    deck_object_key = Column(String(2048), nullable=True)
    source_notes_object_key = Column(String(2048), nullable=True)
    source_notes_html_object_key = Column(String(2048), nullable=True)
    artifact_object_keys = Column(JSONB, nullable=False, default=list)
    share_enabled = Column(Boolean, nullable=False, default=False, index=True)
    share_token_hash = Column(String(64), nullable=True, index=True)
    share_token_nonce = Column(String(64), nullable=True)
    created_at = Column(DateTime, default=_utcnow, nullable=False)
    updated_at = Column(DateTime, default=_utcnow, onupdate=_utcnow, nullable=False)
    deleted_at = Column(DateTime, nullable=True, index=True)

    __table_args__ = (
        Index(
            "uq_learning_decks_user_source_active",
            "user_id",
            "source_identity",
            unique=True,
            postgresql_where=text("deleted_at IS NULL"),
        ),
        Index("idx_learning_decks_user_updated", "user_id", "updated_at"),
    )


class LearningDeckRun(Base):
    """One generation attempt for a Learning Deck."""

    __tablename__ = "learning_deck_runs"

    id = Column(Integer, primary_key=True)
    deck_id = Column(Integer, nullable=False, index=True)
    user_id = Column(Integer, nullable=False, index=True)
    llm_task_id = Column(
        Integer,
        ForeignKey("llm_tasks.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    status = Column(String(32), nullable=False, default="queued", index=True)
    interests_prompt = Column(Text, nullable=True)
    source_snapshot = Column(JSONB, nullable=False, default=dict)
    timeline = Column(JSONB, nullable=False, default=list)
    artifact_storage_prefix = Column(String(2048), nullable=True)
    deck_object_key = Column(String(2048), nullable=True)
    source_notes_object_key = Column(String(2048), nullable=True)
    source_notes_html_object_key = Column(String(2048), nullable=True)
    artifact_object_keys = Column(JSONB, nullable=False, default=list)
    model_provider = Column(String(50), nullable=True)
    model_name = Column(String(100), nullable=True)
    sandbox_provider = Column(String(50), nullable=True)
    sandbox_id = Column(String(255), nullable=True)
    agent_log_object_key = Column(String(2048), nullable=True)
    error_message = Column(Text, nullable=True)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=_utcnow, nullable=False)
    updated_at = Column(DateTime, default=_utcnow, onupdate=_utcnow, nullable=False)

    __table_args__ = (
        Index("idx_learning_deck_runs_deck_created", "deck_id", "created_at"),
        Index("idx_learning_deck_runs_user_status", "user_id", "status"),
        Index(
            "uq_learning_deck_runs_user_active",
            "user_id",
            unique=True,
            postgresql_where=text(ACTIVE_LEARNING_DECK_RUN_STATUS_SQL),
        ),
    )

File: client/newsly/newsly/Models/LearningDeck.swift
//
//  LearningDeck.swift
//  newsly
//

import Foundation

enum LearningDeckSourceKind: Equatable {
    case content
    case githubRepo
    case unknown(String)

    var rawValue: String {
        switch self {
        case .content:
            return "content"
        case .githubRepo:
            return "github_repo"
        case .unknown(let value):
            return value
        }
    }

    init(rawValue: String) {
        switch rawValue {
        case "content":
            self = .content
        case "github_repo":
            self = .githubRepo
        default:
            self = .unknown(rawValue)
        }
    }
}

enum LearningDeckRunStatus: Equatable {
    case queued
    case preparing
    case generating
    case validating
    case publishing
    case completed
    case failed
    case cancelled
    case ready
    case unknown(String)

    var rawValue: String {
        switch self {
        case .queued:
            return "queued"
        case .preparing:
            return "preparing"
        case .generating:
            return "generating"
        case .validating:
            return "validating"
        case .publishing:
            return "publishing"
        case .completed:
            return "completed"
        case .failed:
            return "failed"
        case .cancelled:
            return "cancelled"
        case .ready:
            return "ready"
        case .unknown(let value):
            return value
        }
    }

    var isActive: Bool {
        switch self {
        case .queued, .preparing, .generating, .validating, .publishing:
            return true
        case .completed, .failed, .cancelled, .ready, .unknown:
            return false
        }
    }

    var displayLabel: String {
        switch self {
        case .queued:
            return "Queued"
        case .preparing:
            return "Preparing"
        case .generating:
            return "Generating"
        case .validating:
            return "Checking"
        case .publishing:
            return "Publishing"
        case .completed, .ready:
            return "Ready"
        case .failed:
            return "Failed"
        case .cancelled:
            return "Cancelled"
        case .unknown:
            return "Pending"
        }
    }

    init(rawValue: String) {
        switch rawValue {
        case "queued":
            self = .queued
        case "preparing":
            self = .preparing
        case "generating":
            self = .generating
        case "validating":
            self = .validating
        case "publishing":
            self = .publishing
        case "completed":
            self = .completed
        case "failed":
            self = .failed
        case "cancelled":
            self = .cancelled
        case "ready":
            self = .ready
        default:
            self = .unknown(rawValue)
        }
    }
}

struct LearningDeckTimelineEntry: Identifiable, Equatable {
    let status: LearningDeckRunStatus
    let note: String
    let createdAt: Date

    var id: String { "\(status.rawValue)-\(ServerDate.format(createdAt))-\(note)" }

    init(status: LearningDeckRunStatus, note: String, createdAt: Date) {
        self.status = status
        self.note = note
        self.createdAt = createdAt
    }

}

struct LearningDeckRun: Identifiable, Equatable {
    let id: Int
    let status: LearningDeckRunStatus
    let interestsPrompt: String?
    let timeline: [LearningDeckTimelineEntry]
    let errorMessage: String?
    let startedAt: Date?
    let completedAt: Date?
    let createdAt: Date
    let updatedAt: Date?

    init(
        id: Int,
        status: LearningDeckRunStatus,
        interestsPrompt: String?,
        timeline: [LearningDeckTimelineEntry],
        errorMessage: String?,
        startedAt: Date?,
        completedAt: Date?,
        createdAt: Date,
        updatedAt: Date?
    ) {
        self.id = id
        self.status = status
        self.interestsPrompt = interestsPrompt
        self.timeline = timeline
        self.errorMessage = errorMessage
        self.startedAt = startedAt
        self.completedAt = completedAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

}

struct LearningDeck: Identifiable {
    let id: Int
    let title: String
    let sourceKind: LearningDeckSourceKind
    let sourceURL: String?
    let sourceContentId: Int?
    let sourceTitle: String?
    let sourceMetadata: [String: AnyCodable]
    let status: LearningDeckRunStatus?
    let shareEnabled: Bool
    let viewerAvailable: Bool
    let sourceNotesAvailable: Bool
    let latestSuccessfulRunId: Int?
    let latestRun: LearningDeckRun?
    let createdAt: Date
    let updatedAt: Date?

    init(
        id: Int,
        title: String,
        sourceKind: LearningDeckSourceKind,
        sourceURL: String?,
        sourceContentId: Int?,
        sourceTitle: String?,
        sourceMetadata: [String: AnyCodable],
        status: LearningDeckRunStatus?,
        shareEnabled: Bool,
        viewerAvailable: Bool,
        sourceNotesAvailable: Bool,
        latestSuccessfulRunId: Int?,
        latestRun: LearningDeckRun?,
        createdAt: Date,
        updatedAt: Date?
    ) {
        self.id = id
        self.title = title
        self.sourceKind = sourceKind
        self.sourceURL = sourceURL
        self.sourceContentId = sourceContentId
        self.sourceTitle = sourceTitle
        self.sourceMetadata = sourceMetadata
        self.status = status
        self.shareEnabled = shareEnabled
        self.viewerAvailable = viewerAvailable
        self.sourceNotesAvailable = sourceNotesAvailable
        self.latestSuccessfulRunId = latestSuccessfulRunId
        self.latestRun = latestRun
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    var displayTitle: String {
        nonEmptyTrimmed(title) ?? nonEmptyTrimmed(sourceTitle) ?? "Learning Deck"
    }

    var statusLabel: String {
        if hasActiveLatestRun {
            return latestRun?.status.displayLabel ?? "Pending"
        }
        if viewerAvailable {
            return "Ready"
        }
        return status?.displayLabel ?? "Pending"
    }

    var hasActiveLatestRun: Bool {
        latestRun?.status.isActive == true
    }

    var latestNote: String? {
        if let errorMessage = nonEmptyTrimmed(latestRun?.errorMessage) {
            return errorMessage
        }
        return latestRun?.timeline.last?.note
    }
}

struct LearningDeckListResponse {
    let decks: [LearningDeck]
}

struct LearningDeckShareResponse {
    let shareEnabled: Bool
    let shareURL: String?
}
